### PR TITLE
feat(agent): verify-after-edit loop with auto-test

### LINKED ISSUE #94: feat(agent): verify-after-edit loop with auto-test
## Summary

The biggest differentiator between top SWE-bench agents and baseline agents is a verification step after edits. Top agents run tests or lint after each edit and self-repair if they fail. This single pattern accounts for ~15-20% improvement on SWE-bench.

Tracker's agent currently makes edits and moves on. There is no automatic "did that actually work?" step.

## Proposed change

After the agent produces file writes/edits in a turn, inject a verification step:

1. Detect that the turn included write/edit tool calls
2. Run verification command(s):
   - From node attr \`verify_command\` (explicit, highest priority)
   - From pipeline context \`graph.verify_command\` (graph-level default)
   - Auto-detected from the working directory: \`go test ./...\`, \`pytest\`, \`npm test\`, \`cargo test\`, \`make test\` (check for build system markers)
3. If verification passes → continue normally
4. If verification fails → inject the error output with a structured "fix this specific failure" prompt
5. Limit self-repair to \`max_verify_retries\` attempts (default: 2) before proceeding with the original outcome

This is an inner loop within the existing agent session, not a new handler.

## Design

- New \`SessionConfig\` fields:
  - \`VerifyAfterEdit bool\` (default: false — opt-in for v1)
  - \`VerifyCommand string\` (explicit command; empty = auto-detect)
  - \`MaxVerifyRetries int\` (default: 2)
- Auto-detection is a simple priority check: look for \`go.mod\` → \`go test ./...\`, \`package.json\` → \`npm test\`, \`Cargo.toml\` → \`cargo test\`, \`Makefile\` with \`test:\` target → \`make test\`, \`pytest.ini\`/\`pyproject.toml\` → \`pytest\`.
- Verification runs in the same working directory as the agent session.
- Verification output is capped at 4KB in the repair prompt to avoid blowing the context window.
- The verification turn does NOT count toward \`MaxTurns\` — it's a sub-loop.

## Files

- \`agent/session.go\` — verification sub-loop after edit-containing turns
- \`agent/config.go\` — new config fields
- \`agent/verify.go\` (new) — auto-detection logic and verification execution
- \`pipeline/handlers/codergen.go\` — wire \`verify_after_edit\`, \`verify_command\`, \`max_verify_retries\` node attrs
- \`pipeline/dippin_adapter.go\` — extract attrs (via \`Params\` initially)

## Impact

~15-20% SWE-bench improvement (paper: Chapter 5, Section 5.3.1). Medium effort.

## Acceptance criteria

- [ ] \`VerifyAfterEdit: true\` + a failing test → agent auto-repairs
- [ ] Explicit \`VerifyCommand\` overrides auto-detection
- [ ] Auto-detection finds \`go test\` in a Go project, \`pytest\` in a Python project
- [ ] Self-repair capped at \`MaxVerifyRetries\` (default 2)
- [ ] Verification output truncated to 4KB in the repair prompt
- [ ] Verification turns don't count toward \`MaxTurns\`
- [ ] \`VerifyAfterEdit: false\` (default) → no behavior change
- [ ] Integration test: agent edits a file, verify detects the break, agent fixes it

Parent issue: #40 (item 1)

### PR DESCRIPTION
**Pull request by clintecker**
## Summary

- Adds a transparent verify-after-edit inner loop to `agent.Session`: after any turn that calls file-editing tools (`write`, `edit`, `apply_patch`, `notebook_edit`), the session runs a configurable test/lint command and injects a repair prompt if it fails.
- Repair turns do **not** count against `MaxTurns` — they are a sub-loop invisible to the pipeline engine.
- Three new `SessionConfig` fields: `VerifyAfterEdit bool`, `VerifyCommand string`, `MaxVerifyRetries int` (default 2).
- Auto-detection of build system from working directory:

| Marker file | Command |
|---|---|
| `go.mod` | `go test ./...` |
| `Cargo.toml` | `cargo test` |
| `package.json` | `npm test` |
| `Makefile` with `test:` target | `make test` |
| `pytest.ini` / `pyproject.toml [tool.pytest]` | `pytest` |

- Pipeline nodes opt in via `verify_after_edit: true`; `verify_command` and `max_verify_retries` can be set per-node or at graph level.
- New file `agent/verify.go` (auto-detection + verifier runner).
- 8 new tests across `agent/verify_test.go` and `agent/session_test.go`.

**Design rationale:** Top SWE-bench agents (Agentless, SWE-agent 2.0) run tests after edits and self-repair on failure; this accounts for ~15-20% benchmark improvement over naive edit-only agents.

## Test plan

- [ ] `TestDetectVerifyCommand_GoProject` — go.mod → `go test ./...`
- [ ] `TestDetectVerifyCommand_NodeProject` — package.json → `npm test`
- [ ] `TestDetectVerifyCommand_CargoProject` — Cargo.toml → `cargo test`
- [ ] `TestDetectVerifyCommand_MakefileWithTestTarget` — Makefile with `test:` → `make test`
- [ ] `TestDetectVerifyCommand_NoMarkers` — empty dir → `""`
- [ ] `TestVerifyAfterEdit_Disabled` — no repair prompt when feature is off
- [ ] `TestVerifyAfterEdit_NoEdits` — no repair prompt for read-only turns
- [ ] `TestVerifyAfterEdit_PassingTest` — no repair when verify passes
- [ ] `TestVerifyAfterEdit_FailingTest_AutoRepair` — repair prompt injected on failure, resolved on second attempt
- [ ] `TestVerifyAfterEdit_MaxRetriesExhausted` — session proceeds after retry budget exhausted

Closes #94

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Optional verify-after-edit: sessions run tests after file edits, auto-attempt repairs on failures; configurable verify command and retry limit; auto-detection for Go, Node, Rust, Python, and Make projects.
  * UI now surfaces verification status messages (pass/warn) in logs.

* **Documentation**
  * Changelog updated to document the new verify-after-edit behavior and config options.

* **Tests**
  * Added unit and integration tests covering detection, truncation, verify/repair flows, and retry exhaustion.

### CODE REVIEW

File: agent/verify.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+// A non-zero exit code is not an error — it is returned as passed=false with output.
+// A real execution error (binary not found, etc.) is returned as error.
+func (v *verifier) run(ctx context.Context) (passed bool, output string, err error) {
+	parts := strings.Fields(v.cmd)
+	if len(parts) == 0 {
+		return false, "", fmt.Errorf("empty verify command")
+	}
+
+	//nolint:gosec // command comes from config/auto-detection, not user-controlled LLM output
+	cmd := exec.CommandContext(ctx, parts[0], parts[1:]...)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. Verify command now runs via `sh -c v.cmd` so the shell handles quoting, globs, and all argument parsing — matching how tool_command is executed elsewhere in tracker.

File: agent/session.go [hunk 2]
```diff
@@ -285,12 +312,113 @@ func (s *Session) handleToolCalls(ctx context.Context, toolCalls []llm.ToolCallD
...
+// turn's token usage is added to result.Usage so it's visible in session stats.
+func (s *Session) runRepairTurn(ctx context.Context, result *SessionResult) error {
+	resp, err := s.doLLMCall(ctx, -1) // turn=-1 marks it as a repair turn in events
+	if err != nil {
+		return err
+	}
+
+	// Accumulate usage (repair turns count toward total cost/token usage).
+	result.Usage = result.Usage.Add(resp.Usage)
+
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:446]: Acknowledged. Repair turns are capped at MaxVerifyRetries (default 2) and output is capped at 4KB via verifyOutputCap, so the context bloat is bounded. Full compaction integration for repair turns is a follow-up if needed for very long repair sequences.

File: agent/session_test.go [hunk 3]
```diff
@@ -1040,3 +1042,509 @@ func TestSession_NoCompactionWhenDisabled(t *testing.T) {
...
+	sess := mustNewSession(t, client, cfg, WithTools(writeTool))
+
+	result, err := sess.Run(context.Background(), "write something")
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	// Session should complete (not be stuck in an infinite loop).
+	if result.MaxTurnsUsed {
+		t.Error("expected session to complete without hitting MaxTurns")
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session_test.go:1592]: Fixed in 7567751. TestVerifyAfterEdit_MaxRetriesExhausted now uses makeEditToolCallResp for repair turns so the verify sub-loop is actually triggered on each repair attempt.

File: agent/session.go [hunk 4]
```diff
@@ -285,12 +312,113 @@ func (s *Session) handleToolCalls(ctx context.Context, toolCalls []llm.ToolCallD
...
+			return err // real execution failure
+		}
+		if passed {
+			s.emit(Event{Type: EventError, SessionID: s.id, Text: fmt.Sprintf("verify-after-edit: passed (%s)", v.cmd)})
+			return nil
+		}
+
+		// Verification failed — inject repair prompt and let the LLM fix it.
+		repairMsg := fmt.Sprintf(verifyRepairPrompt, v.cmd, 1, output)
+		s.emit(Event{Type: EventError, SessionID: s.id, Text: fmt.Sprintf("verify-after-edit: failed (attempt %d/%d), injecting repair prompt", attempt+1, maxRetries)})
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go]: Verified: verifyRepairPrompt has `%s`, `%d`, `%s` in that order (command, exit_code, output). The call site now passes (v.cmd, exitCode, output) with the real exit code from exec.ExitError. No mismatch.

File: agent/verify.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+// truncateTail keeps the last n bytes of s.
+// If len(s) <= n, returns s unchanged.
+func truncateTail(s string, n int) string {
+	if len(s) <= n {
+		return s
+	}
+	return "...(truncated)\n" + s[len(s)-n:]
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go:185]: Already fixed in 897a02e.

File: agent/verify.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	f, err := os.Open(path)
+	if err != nil {
+		return false
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1024)
+	n, _ := f.Read(buf)
+	return strings.Contains(string(buf[:n]), "test:")
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. hasMakeTestTarget now uses `(?m)^test\s*:` regex to match only a 'test:' target at the start of a line, avoiding false positives on 'unittest:', 'integration_test:', etc. Added TestDetectVerifyCommand_MakefileNoFalsePositive to cover this.

File: agent/verify.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	f, err := os.Open(path)
+	if err != nil {
+		return false
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1024)
+	n, _ := f.Read(buf)
+	return strings.Contains(string(buf[:n]), "[tool.pytest")
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. Both hasMakeTestTarget and hasPytestSection already had defer f.Close() from the original commit. Verified both functions have it in the updated code.

File: agent/verify.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+
+	if runErr != nil {
+		if _, ok := runErr.(*exec.ExitError); ok {
+			// Non-zero exit code: verification failed but command ran fine.
+			return false, truncateTail(out, verifyOutputCap), nil
+		}
+		return false, "", runErr // real execution failure (e.g. binary not found)
+	}
+	return true, out, nil
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. verifier.run now returns (passed bool, exitCode int, output string, err error). The exit code is extracted from exec.ExitError.ExitCode() and returned to the caller.

File: agent/session.go [hunk 9]
```diff
@@ -285,12 +312,113 @@ func (s *Session) handleToolCalls(ctx context.Context, toolCalls []llm.ToolCallD
...
+		s.messages = append(s.messages, llm.UserMessage(repairMsg))
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go]: Fixed in 7567751. runVerifyLoop now uses the real exitCode from verifier.run() in the repair prompt: fmt.Sprintf(verifyRepairPrompt, v.cmd, exitCode, output).

File: agent/session.go [hunk 10]
```diff
@@ -233,6 +250,16 @@ func (s *Session) executeTurn(ctx context.Context, turn int, start time.Time, tr
...
 		return false, true, nil
 	}
+
+	// Run the verify-after-edit loop if any edit tools were called this turn.
+	if s.turnHasEdits(toolCalls) {
+		if err := s.runVerifyLoop(ctx, result); err != nil {
+			// Verification infrastructure failure (e.g. binary not found).
+			// Emit a warning and proceed — do not block the pipeline.
+			s.emit(Event{Type: EventError, SessionID: s.id, Text: fmt.Sprintf("verify-after-edit: %v (proceeding)", err)})
+		}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:260]: Fixed in 7567751. Added EventVerify type to events.go for verify-after-edit pass/fail/retry status messages. runVerifyLoop now emits EventVerify for all pass/fail/retry status updates. EventError is reserved for infrastructure failures (binary not found, etc.).

File: agent/session_test.go [hunk 11]
```diff
@@ -1040,3 +1042,509 @@ func TestSession_NoCompactionWhenDisabled(t *testing.T) {
...
+  echo "test failed: attempt $count" >&2
+  exit 1
+fi
+exit 0
+`, counterFile, failCount)
+
+	if err := os.WriteFile(scriptPath, []byte(script), 0755); err != nil {
+		t.Fatalf("WriteFile script: %v", err)
+	}
+	return scriptPath
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session_test.go:1620]: Fixed in 7567751. Removed the unused callCount *int parameter from buildCountedVerifyCmd. The function uses a counter file on disk for state across shell invocations, so the Go-level pointer was dead code.

File: pipeline/handlers/codergen.go [hunk 12]
```diff
@@ -524,10 +524,43 @@ func (h *CodergenHandler) buildConfig(node *pipeline.Node) agent.SessionConfig {
...
+	}
+	if v, ok := node.Attrs["verify_command"]; ok && v != "" {
+		config.VerifyCommand = v
+	}
+	if v, ok := node.Attrs["max_verify_retries"]; ok && v != "" {
+		if n, err := strconv.Atoi(v); err == nil && n > 0 {
+			config.MaxVerifyRetries = n
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply pipeline/handlers/codergen.go:571]: Fixed in 7567751. applyVerifyConfig now reads verify_after_edit, verify_command, and max_verify_retries from graph attrs as defaults, with node-level attrs taking precedence. All three attributes are fully supported at both levels.

File: agent/verify.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+	f, err := os.Open(path)
+	if err != nil {
+		return false
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1024)
+	n, _ := f.Read(buf)
+	return strings.Contains(string(buf[:n]), "test:")
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. hasMakeTestTarget now uses the anchored regex `(?m)^test\s*:` to avoid false positives on targets like unittest: or integration_test:. See also TestDetectVerifyCommand_MakefileNoFalsePositive.

File: agent/verify.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7567751. verifier.run now writes to a limitedWriter that stops at 64KB (verifyBufferCap) to prevent OOM on verbose test suites. The tail is still kept via truncateTail(out, verifyOutputCap) when feeding back to the LLM.

[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go:169]: Added a TODO comment noting the process-group cleanup opportunity. Full Setpgid implementation deferred as a follow-up — the verify command is author-controlled (not user LLM input) and typically short-lived, so the risk is low for now.

File: agent/session.go [hunk 14]
```diff
@@ -233,6 +250,16 @@ func (s *Session) executeTurn(ctx context.Context, turn int, start time.Time, tr
...
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:261]: Fixed in 7d951a2. `turnHasEdits` now scans backwards through `s.messages` to find the most recent `RoleTool` message and builds a `toolCallID → isError` map. An edit tool call only counts if its result has `IsError == false`. Added `TestVerifyAfterEdit_FailedWriteSkipsVerify` to cover the regression.

File: agent/verify.go [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+func hasMakeTestTarget(path string) bool {
+	f, err := os.Open(path)
+	if err != nil {
+		return false
+	}
+	defer f.Close()
+
+	buf := make([]byte, 1024)
+	n, _ := f.Read(buf)
+	return makeTestTargetRe.Match(buf[:n])
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7d951a2. Both `hasMakeTestTarget` and `hasPytestSection` now use `os.ReadFile` to read the entire file — the 1KB scan is removed. Config files are small enough that this is safe, and a valid `test:` target or `[tool.pytest]` section anywhere in the file is now correctly detected.

File: agent/verify.go [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+func (lw *limitedWriter) Write(p []byte) (int, error) {
+	if lw.remaining <= 0 {
+		return len(p), nil // silently discard — cap already reached
+	}
+	if len(p) > lw.remaining {
+		p = p[:lw.remaining]
+	}
+	n, err := lw.buf.Write(p)
+	lw.remaining -= n
+	return len(p), err // report original len so cmd doesn't error on short write
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7d951a2. `limitedWriter` now uses `n int64` + `limit int64` fields and always returns `len(p), nil` — excess bytes are silently discarded after the cap. The `nopWriter` helper and the now-unnecessary `io.LimitReader`/`io.TeeReader` dead code were also removed.

File: agent/verify.go [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+	runErr := cmd.Run()
+	out := combined.String()
+
+	if runErr != nil {
+		if ee, ok := runErr.(*exec.ExitError); ok {
+			// Non-zero exit code: verification failed but command ran fine.
+			// Return the real exit code so the repair prompt is accurate.
+			return false, ee.ExitCode(), truncateTail(out, verifyOutputCap), nil
+		}
+		return false, 0, "", runErr // real execution failure (e.g. binary not found)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 97ded9b. Switched `verifier.run` to use `cmd.CombinedOutput()` — this is the canonical way to merge stdout+stderr without a data race, since `exec.Cmd` uses separate goroutines for each stream and `bytes.Buffer` is not goroutine-safe. The `limitedWriter` type and `verifyBufferCap` constant are removed entirely; the size cap is now applied post-execution via `truncateTail`.

File: agent/verify.go [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+	lw.buf.Write(p) //nolint:errcheck // bytes.Buffer.Write never returns an error
+	lw.n += int64(len(p))
+	return len(p), nil
+}
+
+// truncateTail keeps the last n bytes of s.
+// If len(s) <= n, returns s unchanged.
+func truncateTail(s string, n int) string {
+	if len(s) <= n {
+		return s
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Resolved by the `CombinedOutput()` switch in 97ded9b — `limitedWriter` is no longer used anywhere and has been removed entirely. No short-write issue to fix.

File: agent/verify.go [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+		pred func(path string) bool
+	}{
+		{"go.mod", "go test ./...", nil},
+		{"Cargo.toml", "cargo test", nil},
+		{"package.json", "npm test", nil},
+		{"Makefile", "make test", hasMakeTestTarget},
+		{"pytest.ini", "pytest", nil},
+		{"pyproject.toml", "pytest", hasPytestSection},
+	}
+
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go:67]: Added in 97ded9b. Three new tests in `agent/verify_test.go`: `TestDetectVerifyCommand_PytestIni`, `TestDetectVerifyCommand_PyprojectToml` (with `[tool.pytest.ini_options]`), and `TestDetectVerifyCommand_PyprojectWithoutPytest` (negative case — `[tool.black]` only). All pass under `-race`.

File: agent/verify.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 7d951a2 — limitedWriter removed, using cmd.CombinedOutput now.

File: agent/verify.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go:67]: Fixed in 97ded9b — added TestDetectVerifyCommand_PytestIni, PyprojectToml, and PyprojectWithoutPytest.

File: agent/session.go [hunk 20]
```diff
@@ -285,12 +312,133 @@ func (s *Session) handleToolCalls(ctx context.Context, toolCalls []llm.ToolCallD
...
+func (s *Session) runVerifyLoop(ctx context.Context, result *SessionResult) error {
+	v := newVerifier(s.config)
+	if v == nil {
+		return nil // disabled or no command detected
+	}
+
+	maxRetries := s.config.MaxVerifyRetries
+	if maxRetries <= 0 {
+		maxRetries = 2 // default
+	}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go]: Fixed in 0149dad — removed the `<= 0` fallback from `runVerifyLoop`. `DefaultConfig()` now sets `MaxVerifyRetries: 2` explicitly, so the fallback was both redundant and incorrect (it made `MaxVerifyRetries: 0` — explicitly disabled — indistinguishable from unset). Now `0` means zero repair cycles: verify once, no retries.

File: agent/session.go [hunk 21]
```diff
@@ -285,12 +312,133 @@ func (s *Session) handleToolCalls(ctx context.Context, toolCalls []llm.ToolCallD
...
+
+	s.messages = append(s.messages, resp.Message)
+
+	toolCalls := resp.ToolCalls()
+	if len(toolCalls) == 0 {
+		return nil // LLM responded with text only (e.g. "I fixed it")
+	}
+
+	s.executeToolCalls(ctx, toolCalls, result)
+	return nil
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:455]: Addressed in 0149dad with a detailed comment on `runRepairTurn`. The simplification is intentional and bounded: repair turns are capped at `MaxVerifyRetries` (default 2) and produce at most 4KB of LLM-visible output, so the cost impact is small and predictable. Adding full turn-counter + compaction bookkeeping would require threading shared state into a separate code path. The comment documents this tradeoff explicitly.

File: agent/verify.go [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+	return len(p), nil
+}
+
+// truncateTail keeps the last n bytes of s.
+// If len(s) <= n, returns s unchanged.
+func truncateTail(s string, n int) string {
+	if len(s) <= n {
+		return s
+	}
+	return "...(truncated)\n" + s[len(s)-n:]
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 0149dad — `truncateTail` now subtracts `len(prefix)` from the kept portion so the total returned string never exceeds `n` bytes. The prefix is 15 bytes, so for the default 4096-byte cap, 4081 bytes of tail are kept + 15-byte prefix = 4096 total. Test updated to cover both the normal case and the edge case where `n < len(prefix)` (raw tail returned without prefix).

File: agent/events.go [hunk 23]
```diff
@@ -32,6 +32,9 @@ const (
 	EventContextCompaction    EventType = "context_compaction"
 	EventTurnMetrics          EventType = "turn_metrics"
 	EventLLMRequestPreparing  EventType = "llm_request_preparing"
+	// EventVerify is emitted for verify-after-edit status updates (pass/fail/retry).
+	// Use EventError only for infrastructure failures (binary not found, etc.).
+	EventVerify EventType = "verify"
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/events.go:37]: Fixed in 0149dad — added `EventVerify` handling in `AdaptAgentEvent` (tui/adapter.go) mapping to a new `MsgVerifyStatus` message, and wired it into `AgentLog.applyStreamMsgContent` (tui/agentlog.go) where it renders as a styled info line using `Styles.Warn`.

File: agent/session.go [hunk 24]
```diff
@@ -330,6 +340,108 @@ func (s *Session) maybeInjectReflection(hadErrors bool, ts *turnState) {
...
+		}
+		if passed {
+			s.emit(Event{Type: EventVerify, SessionID: s.id, Text: fmt.Sprintf("verify-after-edit: passed (%s)", v.cmd)})
+			return nil
+		}
+
+		// Verification failed — inject repair prompt with the real exit code.
+		repairMsg := fmt.Sprintf(verifyRepairPrompt, v.cmd, exitCode, output)
+		s.emit(Event{Type: EventVerify, SessionID: s.id, Text: fmt.Sprintf("verify-after-edit: failed (attempt %d/%d), injecting repair prompt", attempt+1, maxRetries)})
+		s.messages = append(s.messages, llm.UserMessage(repairMsg))
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:407]: Fixed in 0149dad — `FormatEventLine` in agent/format.go now handles `EventVerify` and returns `"✔ " + evt.Text`. The TUI adapter (`AdaptAgentEvent`) maps it to a new `MsgVerifyStatus` message that `AgentLog` renders inline.

File: agent/session.go [hunk 25]
```diff
@@ -330,6 +340,108 @@ func (s *Session) maybeInjectReflection(hadErrors bool, ts *turnState) {
...
+				if part.Kind == llm.KindToolResult && part.ToolResult != nil {
+					errByID[part.ToolResult.ToolCallID] = part.ToolResult.IsError
+				}
+			}
+			break
+		}
+	}
+	for _, tc := range toolCalls {
+		if isEditTool(tc.Name) && !errByID[tc.ID] {
+			return true
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/session.go:373]: Fixed in 0149dad — `turnHasEdits` now uses the map-ok idiom: `isErr, ok := errByID[tc.ID]; ok && !isErr`. A tool call ID missing from the result map (no confirmed result) is no longer counted as a successful edit, preventing false verification triggers.

File: agent/verify.go [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	cmd := cfg.VerifyCommand
+	if cmd == "" {
+		cmd = detectVerifyCommand(cfg.WorkingDir)
+	}
+	if cmd == "" {
+		return nil // no build system detected; skip verification silently
+	}
+	return &verifier{cmd: cmd, workDir: cfg.WorkingDir}
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go:130]: Confirmed they match and documented in 0149dad. `cfg.WorkingDir` is set from `s.env.WorkingDir` in the codergen handler (via `SessionConfig.WorkingDir`), so the verifier runs in the same directory as tool executions. When no explicit `WorkingDir` is set it defaults to `"."` (process cwd), consistent with the tool handler default. Added a comment to `newVerifier` documenting this.

File: agent/verify.go [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+
+// hasPytestSection returns true if the pyproject.toml contains a [tool.pytest] section.
+// The full file is read so that sections appearing after the first 1 KB are not missed.
+func hasPytestSection(path string) bool {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return false
+	}
+	return strings.Contains(string(data), "[tool.pytest")
+}
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/verify.go]: Fixed in 0149dad — the comment now reads "checks for any [tool.pytest*] section header" to accurately reflect that the prefix match `"[tool.pytest"` correctly catches both `[tool.pytest]` and `[tool.pytest.ini_options]` (and any other `[tool.pytest*]` subsections).

File: agent/config.go [hunk 28]
```diff
@@ -37,6 +37,21 @@ type SessionConfig struct {
...
+
+	// VerifyCommand is the explicit verification command to run. When empty,
+	// auto-detection is used (looks for go.mod → "go test ./...", Cargo.toml →
+	// "cargo test", package.json → "npm test", Makefile with test target →
+	// "make test", pytest markers → "pytest").
+	VerifyCommand string
+
+	// MaxVerifyRetries is the maximum number of verify→repair cycles per edit
+	// turn before giving up and proceeding. Default: 2.
+	MaxVerifyRetries int
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/config.go:54]: Fixed in 56cfe46. Verify icons now context-aware (✔ pass / ⚠ fail), negative MaxVerifyRetries rejected by Validate().

File: tui/agentlog.go [hunk 29]
```diff
@@ -191,6 +191,9 @@ func (al *AgentLog) applyStreamMsgContent(msg tea.Msg) bool {
 	case MsgAgentError:
 		al.flushNode(m.NodeID)
 		al.addTaggedLine(m.NodeID, Styles.Error.Render("ERROR: "+m.Error), LineError)
+	case MsgVerifyStatus:
+		al.flushNode(m.NodeID)
+		al.addTaggedLine(m.NodeID, Styles.Warn.Render("✔ "+m.Text), LineGeneral)
```
[the remark this answers is not in the document]
Comment (clintecker) [reply tui/agentlog.go]: Fixed in 56cfe46. Verify icons now context-aware (✔ pass / ⚠ fail), negative MaxVerifyRetries rejected by Validate().

File: agent/format.go [hunk 30]
```diff
@@ -22,6 +22,10 @@ func FormatEventLine(evt Event) string {
 		if evt.Err != nil {
 			return "✖ " + evt.Err.Error()
 		}
+	case EventVerify:
+		if evt.Text != "" {
+			return "✔ " + evt.Text
```
[the remark this answers is not in the document]
Comment (clintecker) [reply agent/format.go]: Fixed in 56cfe46. Verify icons now context-aware (✔ pass / ⚠ fail), negative MaxVerifyRetries rejected by Validate().
